File: mobile/src/db/seed.ts
import { ExpoSQLiteDatabase } from 'drizzle-orm/expo-sqlite';
import { count } from 'drizzle-orm';
import * as schema from './schema';
import deckData from '../../data/decks.json';

type DeckJson = {
  name: string;
  isFree: boolean;
  productId: string;
  image: string;
  priority: number;
  cards: { key: number; category: string; items: string[]; fact: string }[];
};

import { slugify } from '../utils/slugify';

export async function seedDecks(db: ExpoSQLiteDatabase<typeof schema>) {
  const [{ total }] = await db.select({ total: count() }).from(schema.decks);
  if (total > 0) return;

  const decks = deckData.decks as DeckJson[];

  await db.transaction(async (tx) => {
    for (const deck of decks) {
      const deckId = slugify(deck.name);

      await tx.insert(schema.decks).values({
        id: deckId,
        name: deck.name,
        priority: deck.priority,
        image: deck.image,
        isFree: deck.isFree,
        productId: deck.productId,
      });

      for (const card of deck.cards) {
        await tx.insert(schema.cards).values({
          id: card.key,
          category: card.category,
          items: JSON.stringify(card.items),
          fact: card.fact,
          deckId,
        });
      }
    }
  });

  console.log(`Seeded ${decks.length} decks`);
}

File: api/seed.py
"""Seed the database from data/decks.json."""

import json
import re
from pathlib import Path

from sqlmodel import Session, select

from api.db import engine
from api.models.deck import Card, CardItem, Deck


def slugify(name: str) -> str:
    """Port of mobile/src/utils/slugify.ts — must produce identical IDs."""
    slug = re.sub(r"[^a-z0-9]+", "-", name.lower())
    return slug.strip("-")


def seed_decks() -> None:
    data_path = Path(__file__).resolve().parent.parent / "data" / "decks.json"
    with open(data_path) as f:
        data = json.load(f)

    with Session(engine) as session:
        if session.exec(select(Deck)).first():
            print("Decks already seeded, skipping.")
            return

        deck_count = 0
        card_count = 0
        item_count = 0

        for deck_json in data["decks"]:
            deck_id = slugify(deck_json["name"])
            deck = Deck(
                id=deck_id,
                name=deck_json["name"],
                priority=deck_json["priority"],
                image=deck_json["image"],
                is_free=deck_json["isFree"],
                product_id=deck_json.get("productId"),
            )
            session.add(deck)
            deck_count += 1

            for card_json in deck_json["cards"]:
                card = Card(
                    id=card_json["key"],
                    category=card_json["category"],
                    fact=card_json.get("fact"),
                    deck_id=deck_id,
                )
                session.add(card)
                card_count += 1

                for position, item_text in enumerate(card_json["items"]):
                    item = CardItem(
                        text=item_text,
                        position=position,
                        is_active=True,
                        card_id=card_json["key"],
                    )
                    session.add(item)
                    item_count += 1

        session.commit()
        print(f"Seeded {deck_count} decks, {card_count} cards, {item_count} items.")


if __name__ == "__main__":
    seed_decks()
